peps/pep-0780.rst

```diff
@@ -0,0 +1,447 @@
+
+
+Addition to the Python Standard Library
+'''''''''''''''''''''''''''''''''''''''
+
+Making the ABI features available in an easily accessible, expressive,
+standardized way is useful beyond the scope of environment markers.
+For example, ``"32-bit" in sys.abi_features`` is much more expressive than the
+current standard test of comparing ``sys.maxsize`` with  ``2**32``, which can
+be found more than ten thousand times on GitHub.
+If one wants to determine whether the interpreter is a debug build, there is
+currently no standardized, cross platform way to do so.
+Hence, the `ABI Features`_ listed above are added to the Python standard
+library.
+
+Since they are all the result of compile time choices describing basic features
+of the interpreter, the most intuitive place to put them is in ``sys``.
+Since there is no intrinsic order, nor a possibility for duplication, they are
+added as a ``frozenset`` of strings.
+
+All Python interpreters MUST provide the ``sys.abi_features`` attribute as a
+``frozenset`` of strings, which MUST contain only the `ABI Features`_ that are
+defined in this PEP or in a subsequent PEP.
+
+An example value would be ``sys.abi_features == frozenset({"free-threading",
+"debug", "32-bit"})`` on a free-threaded debug build for win32.
+
+Environment Markers
+'''''''''''''''''''
+
+To make ABI features available in dependency specifications, a new environment
+marker variable, ``sys_abi_features``, is added to the format of dependency
+specifiers with the same semantics as the ``sys.abi_features`` attribute
+proposed above.
+
+To do this, we need to extend the grammar laid out in :pep:`508` and maintained
+in the :ref:`packaging:dependency-specifiers` and document the possible values.
+
+The grammar is extended to include the ``sys_abi_features`` marker variable by
+augmenting the definition of ``env_var`` as follows::
+
+    env_var       = ('python_version' | 'python_full_version' |
+                     'os_name' | 'sys_platform' | 'platform_release' |
+                     'platform_system' | 'platform_version' |
+                     'platform_machine' | 'platform_python_implementation' |
+                     'implementation_name' | 'implementation_version' |
+                     'sys_abi_features' |
+                     'extra' # ONLY when defined by a containing layer
+                     )
+
+Like the grammar, also the overview table of environment markers in
+:ref:`packaging:dependency-specifiers` is augmented to add the following row:
+
+.. list-table::
+    :header-rows: 1
+
+    * - Marker
+      - Python equivalent
+      - Sample values
+    * - ``sys_abi_features``
+      - ``sys.abi_features`` [#sys-abi-features]_
+      - ``set()``, ``{"free-threading"}``, ``{"free-threading", "debug"}``
+
+With these additions, ABI features can be used in dependency specifications via
+the ``in`` operator to test for the presence of a feature, or the ``not in``
+operator to test for the absence of a feature.
```

Comment:
Could you add a sentence about how bad comparisons such as `"3.10.0" < "sys_abi_features"` should be treated? (As someone who has to support such checks already for other environment markers such as `extra`, I'd prefer to both allow to both error and allow to evaluate to false, at the implementations choice)

Reply:
I'm happy if we agree that the marker introduced by this PEP uses the same semantics for set-based markers as defined by PEP 751.